tests/discovery/processing/test_discovery_service_run_experiment.py

```diff
@@ -0,0 +1,150 @@
+
+    assert df_with_prediction.replace({np.nan: None}).to_dict() == TEST_RF_WITHOUT_THRESH_PRED
+    assert mock_save_prediction_called_with.dataset_used_for_prediction == 'test_data'
+    assert mock_save_prediction_called_with.metadata == TEST_RF_WITHOUT_THRESH_CONFIG
+    assert mock_save_prediction_called_with.dataframe.replace(
+        {np.nan: None}).to_dict() == TEST_RF_WITHOUT_THRESH_PRED
+    assert scatter_plot == 'Dummy Plot'
+
+    assert mock_save_tsne_plot_data_called_with.utility.replace({np.nan: None}).to_dict() == TEST_RF_TSNE_PLOT_UTILITY
+    assert mock_save_tsne_plot_data_called_with.features_df.replace(
+        {np.nan: None}).to_dict() == TEST_TSNE_PLOT_FEATURES_INDEX
+    assert list(mock_save_tsne_plot_data_called_with.index_all_labelled.values) == [0, 1, 2, 3]
+    assert list(mock_save_tsne_plot_data_called_with.index_none_labelled.values) == [4, 5, 6, 7, 8, 9, 10, 11, 12]
+
+
+# Since we already check for the data of the tsne plot generation in the tests above,
+# we restrict this test to the experiment data itself and only make sure that the tsne plot save method was called
+def test_run_experiment_with_thresholds_and_gauss_and_saves_result(monkeypatch):
+    mock_save_prediction_called_with = None
+
+    def mock_save_prediction(prediction):
+        nonlocal mock_save_prediction_called_with
+        mock_save_prediction_called_with = prediction
+        return None
+
+    mock_save_tsne_plot_data_called_with = None
+
+    def mock_save_tsne_plot_data(tsne_plot_data):
+        nonlocal mock_save_tsne_plot_data_called_with
+        mock_save_tsne_plot_data_called_with = "Saving TSNE Plot Data"
+        return None
+
+    monkeypatch.setattr(DiscoveryPersistence, 'save_prediction', mock_save_prediction)
+    monkeypatch.setattr(DiscoveryPersistence, 'save_tsne_plot_data', mock_save_tsne_plot_data)
+    _mock_dataset_and_plot(monkeypatch, TEST_GAUSS_WITH_THRESH_INPUT, 'X')
+
+    df_with_prediction, scatter_plot = DiscoveryService.run_experiment('test_data', TEST_GAUSS_WITH_THRESH_CONFIG)
+
+    assert df_with_prediction.replace({np.nan: None}).to_dict() == TEST_GAUSS_WITH_THRESH_PRED
+    assert mock_save_prediction_called_with.dataset_used_for_prediction == 'test_data'
+    assert mock_save_prediction_called_with.metadata == TEST_GAUSS_WITH_THRESH_CONFIG
+    assert mock_save_prediction_called_with.dataframe.replace({np.nan: None}).to_dict() == TEST_GAUSS_WITH_THRESH_PRED
+    assert scatter_plot == 'Dummy Plot'
+    assert mock_save_tsne_plot_data_called_with == 'Saving TSNE Plot Data'
+
+
+def test_run_experiment_with_partially_labelled_data(monkeypatch):
+    mock_save_prediction_called_with = None
+
+    def mock_save_prediction(prediction):
+        nonlocal mock_save_prediction_called_with
+        mock_save_prediction_called_with = prediction
+        return None
+
+    mock_save_tsne_plot_data_called_with = None
```

Comment:
the mock is unused (no assertions) here. Either you can also insert on it or remove the nonlocal var and simply return None in the corresponding mock's method. Both is fine for me as we are alreading testing this mock in other tests.

Reply:
removed the unused mock